prql-compiler/src/sql/gen_query.rs

```diff
@@ -107,8 +107,29 @@ fn translate_select_pipeline(
     let order_by = pipeline.pluck(|t| t.into_sort());
     let takes = pipeline.pluck(|t| t.into_take());
     let is_distinct = pipeline.iter().any(|t| matches!(t, SqlTransform::Distinct));
+    let distinct_ons = pipeline.pluck(|t| t.into_distinct_on());
     let distinct = if is_distinct {
         Some(sql_ast::Distinct::Distinct)
+    } else if !distinct_ons.is_empty() {
+        // FIXME: this "works" but is very hacky — it's using the
+        // `translate_select_item` to translate each id into an Expr — is that
+        // correct?
+        //
+        // Do we know we won't have more than one DistinctOn? We have that for
+        // `Distinct`. But this will panic if we have more than one.
+        // And if there the `SelectItem` is not an `UnnamedExpr`, it'll panic.
```

Comment:
Yes, we know there won't be multiple `Distinct`s, but you also have to inforce that for `DistinctOn`. You can do that here: https://github.com/prql/prql/blob/main/prql-compiler/src/sql/srq/anchor.rs#L342-L353

After that, it is safe to do this:
```
let distinct_on = pipeline.pluck(|t| t.into_distinct_on()).into_iter().next();
```